api/src/main/java/org/openmrs/module/fhir2/api/mappings/ObservationCategoryMap.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.api.mappings;
+
+import javax.validation.constraints.NotNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.Properties;
+
+import com.google.common.collect.ImmutableMap;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ObservationCategoryMap {
+	
+	@Autowired
+	private ResourceLoader resourceLoader;
+	
+	private volatile Map<String, String> classDelegate = null;
+	
+	private volatile Map<String, String> conceptDelegate = null;
+	
+	public String getCategory(@NotNull String conceptClassUuid) {
+		setupDelegates();
+		return classDelegate.getOrDefault(conceptClassUuid, null);
+	}
+	
+	public String getConceptClassUuid(@NotNull String category) {
+		setupDelegates();
+		return conceptDelegate.getOrDefault(category, null);
+	}
+	
+	private void setupDelegates() {
+		if (classDelegate == null) {
+			synchronized (this) {
+				if (classDelegate == null) {
+					Resource resource = resourceLoader.getResource("classpath:observationCategoryMap.properties");
+					Properties properties = new Properties();
+					
+					try (InputStream in = resource.getInputStream()) {
+						properties.load(in);
+					}
+					catch (IOException ignored) {}
+					
+					ImmutableMap.Builder<String, String> builder = ImmutableMap.builderWithExpectedSize(properties.size());
+					for (Map.Entry<Object, Object> entry : properties.entrySet()) {
+						builder.put(entry.getValue().toString(), entry.getKey().toString());
+					}
+					
+					classDelegate = builder.build();
+					
+					builder = ImmutableMap.builderWithExpectedSize(properties.size());
+					for (Map.Entry<Object, Object> entry : properties.entrySet()) {
+						builder.put(entry.getKey().toString(), entry.getValue().toString());
+					}
+					
+					conceptDelegate = builder.build();
```

Comment:
I think this can be achieved using one loop with two declarations of `ImmutableMap.Builder<String, String> builder`. One less loop is better

Reply:
You're absolutely right!